server/src/server.rs

```diff
@@ -0,0 +1,693 @@
+	U: HttpBody + Send + 'static,
+	<U as HttpBody>::Error: Send + Sync + StdError,
+	<U as HttpBody>::Data: Send,
+{
+	/// Start responding to connections requests. This will run on the tokio runtime until the server is stopped.
+	pub fn start(mut self, methods: impl Into<Methods>) -> Result<ServerHandle, Error> {
+		let methods = methods.into().initialize_resources(&self.resources)?;
+		let (stop_tx, stop_rx) = watch::channel(());
+
+		let stop_handle = StopHandle::new(stop_rx);
+
+		match self.cfg.tokio_runtime.take() {
+			Some(rt) => rt.spawn(self.start_inner(methods, stop_handle)),
+			None => tokio::spawn(self.start_inner(methods, stop_handle)),
+		};
+
+		Ok(ServerHandle::new(stop_tx))
+	}
+
+	async fn start_inner(self, methods: Methods, stop_handle: StopHandle) {
+		let max_request_body_size = self.cfg.max_request_body_size;
+		let max_response_body_size = self.cfg.max_response_body_size;
+		let max_log_length = self.cfg.max_log_length;
+		let allow_hosts = self.cfg.allow_hosts;
+		let resources = self.resources;
+		let logger = self.logger;
+		let batch_requests_supported = self.cfg.batch_requests_supported;
+		let id_provider = self.id_provider;
+		let max_subscriptions_per_connection = self.cfg.max_subscriptions_per_connection;
+
+		let mut id: u32 = 0;
+		let connection_guard = ConnectionGuard::new(self.cfg.max_connections as usize);
+		let mut connections = FutureDriver::default();
+		let mut incoming = Monitored::new(Incoming(self.listener), &stop_handle);
+
+		loop {
+			match connections.select_with(&mut incoming).await {
+				Ok((socket, _addr)) => {
+					if let Err(e) = socket.set_nodelay(true) {
+						tracing::warn!("Could not set NODELAY on socket: {:?}", e);
+						continue;
+					}
+
+					let conn = match connection_guard.try_acquire() {
+						Some(conn) => conn,
+						None => {
+							tracing::warn!("Too many connections. Please try again later.");
+							connections.add(http::reject_connection(socket).boxed());
+							continue;
+						}
+					};
+
+					let tower_service = TowerService {
+						inner: ServiceData {
+							remote_addr: socket.peer_addr().ok(),
+							methods: methods.clone(),
+							allow_hosts: allow_hosts.clone(),
+							resources: resources.clone(),
+							max_request_body_size,
+							max_response_body_size,
+							max_log_length,
+							batch_requests_supported,
+							id_provider: id_provider.clone(),
+							ping_interval: self.cfg.ping_interval,
+							stop_handle: stop_handle.clone(),
+							max_subscriptions_per_connection,
+							conn_id: id,
+							logger: logger.clone(),
+							conn: Arc::new(conn),
+						},
+					};
+
+					let service = self.service_builder.service(tower_service);
+
+					// NOTE: this task will finish after the HTTP request have been finished.
+					//
+					// Thus, if it was a Websocket Upgrade request the background task will be spawned separately.
+					let mut stop_handle2 = stop_handle.clone();
```

Comment:
Nit: does it need cloning? it's not being used anywhere elseby the looks of it :)

Is the comment saying that the `async` block below finishes when the http request is read and handled but that ws will lead to task being spawned?

Reply:
it's in a loop so the clone is needed :(

it was just a side-note to previous behavior i.e, the task would live as long as the `ws connection task` was alive but now "this task" finished once the HTTP request is finished but I removed the comment.

this is because the `hyper callback` just a spawns the task directly without access to `FutureDriver`